src/IROperator.h

```diff
@@ -327,6 +327,41 @@ Expr requirement_failed_error(Expr condition, const std::vector<Expr> &args);
 
 Expr memoize_tag_helper(Expr result, const std::vector<Expr> &cache_key_values);
 
+/** Cast a to the same type with twice the bits. */
+Expr widen(Expr a);
+/** Cast a to the same type with half the bits. */
+Expr narrow(Expr a);
+
+/** Compute widen(a) + widen(b). The result is always signed. */
+Expr widening_add(Expr a, Expr b);
+/** Compute widen(a) * widen(b). a and b may have different signedness. */
+Expr widening_mul(Expr a, Expr b);
+/** Compute widen(a) - widen(b). The result is always signed. */
+Expr widening_sub(Expr a, Expr b);
+/** Compute widen(a) << b. */
+Expr widening_shift_left(Expr a, Expr b);
+/** Compute widen(a) >> b. */
+Expr widening_shift_right(Expr a, Expr b);
+
+/** Compute saturating_add(a, (1 >> min(b, 0)) / 2) << b */
```

Comment:
I think it would be worth noting that the "rounding" part only does something if b is negative, in which case this is equal to rounding_shift_left(a, -b). If you're thinking of positive b the idea of this rounding makes no sense. `1 >> something` is not something you'd see in C.

Reply:
> 1 >> something is not something you'd see in C

Well, it's not something you *should* see in C...